### ISSUE TITLE
Interlocked.Exchange should not raise IDISP003

### ISSUE THREAD

**Issue by jeremyVignelles** [DotNetAnalyzers/IDisposableAnalyzers#209]
I am trying to use Interlocked.Exchange to Dispose() my instance and avoid further usages of it.

Here is how I do:

```cs
            var oldValue = Interlocked.Exchange(ref this._disposable, null);
            oldValue?.Dispose();
```

I get a IDISP003 : Dispose previous before reassigning on `ref this._disposable`, but the IDisposable ownership is in fact transfered to oldValue, which must be disposed.

**Comment by JohanLarsson** [comment]
```cs
namespace N
{
    using System;
    using System.IO;
    using System.Threading;

    sealed class C: IDisposable
    {
        private IDisposable disposable = File.OpenRead(string.Empty);

        public void M(string fileName)
        {
            var oldValue = Interlocked.Exchange(ref this.disposable, File.OpenRead(fileName));
            oldValue?.Dispose();
        }

        public void Dispose()
        {
            var oldValue = Interlocked.Exchange(ref this.disposable, null);
            oldValue?.Dispose();
        }
    }
}
```

^ repro for plenty of bugs, currently not handling `Interlocked.Exchange` at all.

**Comment by jeremyVignelles** [comment]
Low prio for me, I've replaced the code with a SemaphoreSlim
